tests/arch/arm/arm_thread_swap/src/arm_thread_arch.c

```diff
@@ -68,25 +68,27 @@ static void load_callee_saved_regs(const _callee_saved_t *regs)
 	__asm__ volatile (
 		"mov r1, r7;\n\t"
 		"mov r0, %0;\n\t"
-		"ldmia r0!, {r4-r7};\n\t"
+		"add r0, #16;\n\t"
 		"ldmia r0!, {r4-r7};\n\t"
 		"mov r8, r4;\n\t"
 		"mov r9, r5;\n\t"
 		"mov r10, r6;\n\t"
 		"mov r11, r7;\n\t"
+		"sub r0, #32;\n\t"
+		"ldmia r0!, {r4-r7};\n\t"
 		"mov r7, r1;\n\t"
-		:
+		: /* no output */
 		: "r" (regs)
-		: "memory"
+		: "memory", "r1", "r0"
```

Comment:
can you expand why it's not needed to have here `r8 / r9 / ...` as well? We are clobbering those AFAICT

Reply:
The clobber list is there for `%0`, which only occurs once early on, moving it to `r0`.  By the time anything happens to `r4-11`, we no longer have to worry about where `%0` was, since it's already been copied into `r0`, which is on the list (although normally GCC would only select r4-11 anyway).

Reply:
the purpose of this function is to load `r4-r11` from a struct; adding these regs to the clobber list would nullify what this function does (because the compiler will try to restore the content of `r4-r11` "clobbered" in this inline asm block).

yes, it is a very dangerous thing to do ... but, with how this test is currently implemented, there is not much we can do.

Reply:
right, but that really depends on the order of instructions and on the code itself, we have to remember ourselves to update the list if we modify that code.

Formally (IMO) we should list all the clobbered registers (after all, that is what the clobber list is for) and that includes `r8 / r9 / ...` as well.

Just a nit, I know.

Reply:
> because the compiler will try to restore the content of r4-r11 "clobbered" in this inline asm block

@stephanosio 

now I'm confused. I'm still scratching my head over the clobber list but according to the documentation `When the compiler selects which registers to use to represent input and output operands, it does not use any of the clobbered registers. As a result, clobbered registers are available for any use in the assembler code. `

Is GCC really supposed to restore the content of the clobbered registers?

Reply:
That's what I understood ... the clobber list is there to say what can safely be used for `%0`, etc. I'm not sure what to put here, though, since we're effectively using all of the general purpose registers that are available (`r4-11`). We're doing the same in the second instance, where at various points are are changing `r6` and `r7`, but those are the only registers we CAN use for `%0`.  

It seems like we have to have some awareness of the code when defining the clobber list. I know `r6` and `r7` are safe in the second instance, even though they are changed later, and we can't block EVERY register that gets changed. I don't like having to be aware of the code as well, but I don't see an alternative?

Reply:
> Is GCC really supposed to restore the content of the clobbered registers?

If the compiler was already using one of the clobbered registers in the same context, it will have to save and restore the clobbered registers before and after, respectively, the inline asm block.

Reply:
> If the compiler was already using one of the clobbered registers in the same context, it will have to save and restore the clobbered registers before and after, respectively, the inline asm block.

Ok, I got it now, thanks.

Reply:
It seems to me like we always need some awareness of our code when defining the clobber list, but it's a bit opaque to me as well.

For this code:

```c
	__asm__ volatile (
		"push {r4-r7};\n\t"
		"mov r4, r8;\n\t"
		"mov r5, r9;\n\t"
		"push {r4, r5};\n\t"
		"mov r4, r10;\n\t"
		"mov r5, r11;\n\t"
		"push {r4, r5};\n\t"
		"push {r0, r1};\n\t"
		"mov r1, r7;\n\t"
		"mov r0, %0;\n\t"
		"add r0, #16;\n\t"
		"ldmia r0!, {r4-r7};\n\t"
		"mov r8, r4;\n\t"
		"mov r9, r5;\n\t"
		"mov r10, r6;\n\t"
		"mov r11, r7;\n\t"
		"sub r0, #32;\n\t"
		"ldmia r0!, {r4-r7};\n\t"
		"mov r7, r1;\n\t"
		"pop {r0, r1};\n\t"
		: /* no output */
		: "r" (&ztest_thread_callee_saved_regs_container)
		: "memory", "r4", "r5", "r0", "r1", "r8", "r9", "r10", "r11"
	);
```

I need to restrict GCC from selecting `r4, r5, r0, r1` ... but with the `add` and `sub` changes, it seems safe to remove `r8-11` since those could be used in this case for the input/output registers, based on where they occur. I'd need to change my clobber list if `%0` occurred later, or was used again as it was in an earlier version of this code.

I may be missing some other nuance of the clobber list, though, in this case where we have inline ASM mixed with C code, which complicates things.

@stephanosio Would `: "memory", "r4", "r5", "r0", "r1"` be correct here in your opinion? Normally GCC wouldn't select `r0-3` anyway, but it seems prudent to put it just in case since we are using those before `%0`.

Reply:
Still trying to wrap my head around this.

If registers hold temporary values or are modified, it seems like they should be on the clobber list. See armclang documentation for another point of view: https://developer.arm.com/documentation/100067/0610/armclang-Inline-Assembler/Inline-assembly-statements-within-a-function/Clobber-list

> "if a register holds a temporary value, include it in the clobber list"

> "Each string is the name of a register that the assembly code potentially modifies, but for which the final value is not important."

But we're changing every register other than `r2` and `r3` here, which have their own reserved meaning.

Not sure I have the same conclusion as you, @stephanosio, of not including `r4-11` here, but what do you feel is the right list in this case to be sure I understand what you're proposing?

Excluding `r4` and `r5` from the list here will cause a hard fault, though, since we use those earlier in a temporary capacity. The problem is we use every register in a temporary/scratch manner, meaning it should be on the list, BUT registers 4-11 also have 'final values' that are important outside the function ... so in theory they should be excluded, but in practice we can't in this case.

Basically, this code is awkward and I don't think there's a right way to define the clobber list, other than being aware of the contents of the code so excluding everything modified before the input value (r0, r1, r4, r5).

Reply:
I feel like the right list here is: `: "memory", "r4", "r5", "r0", "r1"` but let me know if there's no agreement on that.

Reply:
Was just discussing that elsewhere. We do need r4/r5 to be part for now, though, to avoid the hard fault.

However this code has a number of much larger issues. Are you objectionable to merging the current changes in as a solution to #50709, and I can do a rework of this code in a new PR?

Reply:
No objection to the current solution since it does "fix" the issue and this needs to be reworked anyway.

Reply:
If you don't mind approving, I'll figure out a better solution for the larger problems with this code.

Reply:
@microbuilder PTAL at https://github.com/zephyrproject-rtos/zephyr/pull/51366#discussion_r999631890 about the clobber list, I agree with @stephanosio and I propose a tweak to accommodate it. HTH.